text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,238 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Slightly increases the size of the generated binaries. However, the increase is [typically below 1%](https://github.com/rust-lang/rfcs/pull/2801#issuecomment-549184251).
+- Adds more platform-specific code to the build process, which needs to be maintained.
+- Slightly more work need to be performed at compile time. This implies slightly slower compilation.
+  - If the compilation time impact is deemed to be significant, collecting and embedding this data will be disabled by default in debug profile before stabilization. It will be possible to override this default using the per-profile configuration option.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Rationale
+
+- This way version information is *impossible* to misplace. As long as you have the binary, you can recover the info about dependency versions. The importance of this is impossible to overstate. This allows auditing e.g. a Docker container that you did not build yourself, or a server that somebody's set up a year ago and left no audit trail.
+- A malicious actor could lie about the version information. However, doing so requires modifying the binary - and if a malicious actor can do _that,_ you are pwned anyway. So this does not create any additional attack vectors other than exploiting the tool that's recovering the version information, which can be easily sandboxed.
+- Any binary authentication that might be deployed automatically applies to the version information. There is no need to separately authenticate it.
+- Tooling for extracting information from binaries (such as ELF sections) is already readily available, as are zlib decompressors and JSON parsers. It can be extracted and parsed [in 5 lines of Python](https://github.com/rust-secure-code/cargo-auditable/blob/master/PARSING.md), or even with a shell one-liner in a pinch.
+- This enables third parties such as cloud providers to scan your binaries for you. Google Cloud [already provides such a service](https://cloud.google.com/container-registry/docs/get-image-vulnerabilities), Amazon has [an open-source project you can deploy](https://aws.amazon.com/blogs/publicsector/detect-vulnerabilities-in-the-docker-images-in-your-applications/) while Azure [integrates several partner solutions](https://docs.microsoft.com/en-us/azure/security-center/security-center-vulnerability-assessment-recommendations). They do not support this specific format yet, but integration into Trivy was very easy, so adding support will likely be trivial.
+
+## Why enable this by default?
+
+If you have 10 programs on your computer that use a library exploitable over the network, your computer and all your data are vulnerable as long as even **a single program** remains vulnerable.
```

Comment:
```suggestion
If you have 10 programs on your computer that depend on a library exploitable over the network and use that library in a way that exposes it to exploitation, your computer and your data are vulnerable as long as even **a single program** remains vulnerable.
```
If you'll forgive my pedantry, the buggy library might not be exposed to attackers' input, and some of the data on the computer might be unreachable from the

Reply:
It's out of scope for this RFC, but it'd be nice if Rust tooling could ignore vulnerabilities that are not in the call-graph, the same way the Go tooling does: https://go.dev/blog/vuln

Reply:
There is ongoing work to implement this in https://google.github.io/osv-scanner/, see https://github.com/google/osv-scanner/pull/452

https://github.com/rustfoundation/painter might also come in handy